add flux-dump and flux-restore

garlick opened the pull request:
As proposed in #4202, here are dump/restore utilities that walk KVS metadata using the content interface so they can be used when the KVS is not loaded.  They use [libarchive](https://www.libarchive.org/) to support various portable archive formats.

Still playing with these and plan to add tests and man pages but wanted to get this checkpointed.

garlick commented:
Some numbers from my test cluster, FWIW:
- 22094 inactive jobs
- 1234 MB content.sqlite
- 122s  to load job manager
- 102s to load job-list

With flux stopped I ran this script, which took 5m+57s to run
```bash
#!/bin/bash
SQLITE_FILE=/net/frankenfurter/pool/scratch/flux/content.sqlite

reader() {
        ./flux start -o,-Sbroker.rc1_path= \
                -o,-Sbroker.rc3_path=\
                -o,-Scontent.backing-path=$SQLITE_FILE.bak \
                bash -c "\
                        flux module load content-sqlite && \
                        flux dump --checkpoint - \
                "
}

writer() {
        ./flux start -o,-Sbroker.rc1_path= \
                -o,-Sbroker.rc3_path= \
                -o,-Scontent.backing-path=$SQLITE_FILE \
                bash -c "\
                        flux module load content-sqlite && \
                        flux restore -v --checkpoint - \
                "
}

mv $SQLITE_FILE $SQLITE_FILE.bak
reader | writer
```
Afterwards:
- 22094 inactive jobs (no change)
- 731 MB content.sqlite (0.59 x)
- 40s  to load job manager (0.32 x)
- 54s to load job-list (0.52 x)

grondo commented:
> Other things I thought of but felt could be deferred are:

Another idea we had

 * A filter or transform step before restore which would allow - among other things like pruning jobs - restoring to an instance running on a different resource set. This would perhaps allow for easy checkpoint/restore of Flux instances without requiring the instance restart on the same allocation. (We could call it the "everlasting jobstopper")

chu11 reviewed (commented):
just did a skim, overall looks good, didn't notice much out of place.  just a lot of nits / minor things.

garlick commented on doc/man1/flux-restore.rst:
```diff
@@ -0,0 +1,67 @@
+**-h, --help**
+   Summarize available options.
+
+**-v, --verbose**
```
The level is not used so I'll make it boolean.

garlick commented on doc/man1/flux-dump.rst:
```diff
@@ -0,0 +1,120 @@
+**-h, --help**
+   Summarize available options.
+
+**-v, --verbose**
```
Same here

garlick commented on src/common/libkvs/kvs_checkpoint.c:
```diff
@@ -91,45 +91,21 @@ int kvs_checkpoint_lookup_get_rootref (flux_future_t *f, const char **rootref)
-        return -1;
-
-    if (version != 0 && version != 1) {
-        errno = EINVAL;
+                                "timestamp", &ts) < 0)
         return -1;
-    }
```
Sorry I think you are right.

The thought I was having was that we don't want to artificially prevent a system administrator from reverting a flux-core upgrade that happens to raise  the checkpoint version number here, when the timestamp is probably not changing; however, I think that could be more of an issue for deciding when to bump the version number.  It probably shouldn't be bumped at all unless the new version introduces a breaking change.

So I'll fix that, thanks.

garlick commented on doc/man1/flux-dump.rst:
```diff
@@ -0,0 +1,120 @@
+**-v, --verbose**
+   List keys on stderr as they are dumped.
+
+**--checkpoint**
+   Generate snapshot from the latest checkpoint written to the content
+   backing store, instead of from the current KVS root.
```
Dropping the optional arg so the man page will be correct.

garlick commented on doc/man1/flux-restore.rst:
```diff
@@ -0,0 +1,67 @@
+**-v, --verbose**
+   List keys on stderr as they are restored.
+
+**--checkpoint**
```
Dropping the optional arg so the man page will be correct.

garlick commented:
Just pushed an update with changes squashed (hope that's OK):
- made `--verbose` and `--snapshot` boolean options consistent with man page
- changed kvs snapshot helpers to use the `kvs-primary` if the key is NULL since that's all we support now
- restored snapshot object version check
- fixed other typos and whitespace issues called out by @chu11
- improved testing to ensure that "valref" objects are covered, and to count blobs before and after a restore to show GC benefit
- print number of keys and blobs restored on `flux-restore` completion
- fixed a  json memory leak in `flux-restore` found with valgrind
- rebased on current master

garlick commented:
Oops found some formatting errors in the man pages so forced a push to correct those, and while I was in there, make an editing pass for clarity.

garlick commented:
I'm hitting a surrogate error in @chu11's job-archive sqlite query tool that I reused here to count blobs in the sqlite database, only  in the inception builder.  It looks like it may be fixed by @grondo's pr #4212.  Fortuitous!  I'll try rebasing this PR on that one and see how it goes.

grondo commented:
Just trying out this branch on my size=90 full system instance on fluke with 40K+ jobs.

One thing that might be nice to have is a progress meter in `flux dump/restore`. I'm not sure if that is even possible, but a command running for many minutes with no visible output is somewhat disconcerting, and `-v, --verbose` is way too much information. Anything in-between might be helpful, even just a count of data dumped or restored?

garlick commented:
Argh, sorry, just re-pushed again - I had let a fix for #4210 leak onto this branch that was not quite right.
Also rebased on @grondo's latest copy of his branch.

garlick commented:
Looks like #4212 fixed my problem.  Thanks!

I added a periodic count of dumped keys to stderr to both `flux-dump` and `flux-restore`.  It is disabled if a `-q,--quiet` option is specified, or if `-v,--verbose`. 

Once #4212 gets merged, I'll rebase this on master and push the progress stuff too.

I think the coverage here could be better, but only if we simulate bad  KVS metadata in the store.  That's not going to be fun.  My vote is to save that for another day.

garlick commented:
OK, re-pushed after rebasing, and with new progress display (overwrites line on stderr every 100 keys dumped/restored) and  `--quiet` option.

grondo commented:
Nice! The progress display is very nice, IMO.

chu11 commented:
> Since flux-dump generates the dump by interacting directly with the
content store, a dump may be created when the KVS is not loaded, as
long as the --checkpoint option is supplied so it has a starting blobref.
Otherwise, the starting blobref is obtained by requesting the current
root from the KVS.

Saw this in a commit message and made me wonder, would a `--checkpoint-blobref` option to reasonable for `flux -dump`?  Like if a user just saved off the value after calling `flux kvs getroot`?

Was just a thought.

chu11 commented:
Another random thought / idea.  Should the key a user restores to be a symlink?  Or possibly in a different namespace?  I'm not entirely sure if there's a realistic use case.  I'm imagining like a user wants to re-start their instance from a specific point in time?  Just a random brainstorm, feel free to ignore.

chu11 commented on t/t2807-dump-cmd.t:
```diff
@@ -0,0 +1,172 @@
+	flux content dropcache &&
+	flux module remove content-sqlite
+'
+# Small values are contained in dirent rather than separate objects, so expect:
+#  rootdir-v0 + (rootdir-v1 + a + b + x) + rootdir-v2 + rootdir-v3 + rootdir-v4
```
it took me a while to grok this comment and get to how you go to 8.  Perhaps more clear (I think the +x above is a cut & paste error?)

```
rootdir-v0 (initial rootdir) + (rootdir-v1 (add a) + dir a + dir b) + (rootdir-v2 (add y)) + (rootdir-v3 (add x) + valref x) + (rootdir-v4 (add z)) = 8 blobrefs
```

garlick replied:
sorry, those comments weren't updated when the script changed!  I'll update and expand the comment so the casual reader won't be totally mystified!

chu11 commented on t/t2807-dump-cmd.t:
```diff
@@ -0,0 +1,172 @@
+	flux module remove content-sqlite
+'
+# Intermediate rootdir versions are not restored so expect:
+#  rootdir-v4 + a + b + x
```
for consistency to my 0above comment, dir a, dir b, valref x?

chu11 commented on doc/man1/flux-dump.rst:
```diff
@@ -0,0 +1,124 @@
+DESCRIPTION
+===========
+
+The ``flux-dump`` command write a snapshot of the Flux key value store
```
"writes a snapshot"? (plural write)

chu11 commented on doc/man1/flux-restore.rst:
```diff
@@ -0,0 +1,82 @@
+CAVEATS
+=======
+
+The KVS module writes its final root checkpoint to the backing store when
+it is unloaded, replacing whatever was there before.  Therefore, restoring
+the KVS checkpoint from archive with the KVS module loaded and then unloading
+the KVS module causes the restored blobref to be overwritten.  Please ensure
+the KVS module is unloaded when using the *--checkpoint* option.
```
idea: could kvs-restore check that the module isn't loaded?

garlick replied:
Oooh good idea.

chu11 commented on doc/man1/flux-restore.rst:
```diff
@@ -0,0 +1,82 @@
+KVS service is running, or as a checkpoint in the content backing store
+if *--checkpoint* is used, without the KVS running.  One of those two options
+is required.
+
+
```
perhaps want to add (or add an example?) of how dump + restore garbage collects inactive references?  But on other hand it may not be super wise to advertise, as users may abuse it?

garlick replied:
I think it may be best to hold off on that for a bit.  For now, the things that are omitted from the archive are described in `flux-dump(1)`.  If we're successful with using it for garbage collection, we could revisit.

garlick commented:
Thanks for the additional comments @chu11!

On the two non-inlined ideas:

> would a --checkpoint-blobref option to reasonable for flux -dump? Like if a user just saved off the value after calling flux kvs getroot?

I had a `--blobref` option originally so it would be easy to add back in, but it seemed like any gainful use of it would require too much knowledge about how the KVS internals work, e.g. that you'd want a blobref that refers to a directory that you could get from a commit, or other tricky ways.  I thought the more user-friendly option would be `--key` (a directory name), but that was a little harder to implement, or at least I had a hard time figuring it out at the time.   So...maybe a follow on PR if a use case develops?

> Could the user restoring want the key to be a symlink? Or possibly in a different namespace? I'm not entirely sure if there's a realistic use case. I'm imagining like a user wants to re-start their instance from a specific point in time?

It definitely seems like dump/restore from/to a namespace root makes sense. I actually listed that above as something I thought we might want to add later.   I think a symlink can easily be added after the restore, if the idea there is to  add one that could be followed into another namespace from the primary.

Good ideas IMHO, but could be deferred to if/when a usecase develops?

garlick commented:
Pushed some fixups (squashed) for @chu11's comments.

The test for the kvs being loaded was a great idea - I was able to remove the CAVEATS section from `flux-restore(1)` now that it is automatically detected.

garlick commented:
Well something is not right here.  When I dump and restore my system at home, I can no longer run jobs.  I get the following error

```
$ flux mini run hostname
0.475s: job.exception type=exec severity=0 failed to create guest ns: No such file or directory
$ sudo flux dmesg
[snip]
2022-03-15T04:32:12.313805Z kvs.err[0]: content_load_completion: flux_content_load_get: No such file or directory
2022-03-15T04:32:12.317248Z job-exec.info[0]: job-exception: id=2317853860560896: failed to create guest ns: No such file or directory
```
So it looks like there is a dangling blobref there somewhere.

chu11 reviewed (approved):
updates look good, LGTM!

garlick commented:
Thanks @chu11.

Just pushed  some more fixups (squashed) after running down the problem above:
- add work around and test for #4222
- fixed minor typos in job-exec.c, content-sqlite.c
- flush content store after restoring, unless `--no-cache`
- slightly adjust progress output and make `--quiet` really quiet
- cherry-pick @grondo's commit to add libarchive to Dockerfiles and drop temporary `checks/Dockerfile` commit

grondo commented:
Sorry the fedora33 docker image build or push failed for unknown reason. I had to restart the build so it will be up to an hour before that is fixed.
